python/ray/tests/test_object_assign_owner.py

```diff
@@ -0,0 +1,77 @@
+import pytest
+import ray
+import time
+
+
+@pytest.mark.parametrize(
+    "actor_resources",
+    [
+        dict(
+            zip(["owner", "creator", "borrower"], [{
+                f"node{i}": 1
+            } for i in _])) for _ in [
+                [1, 2, 3],  # None of them is on the same node.
+                [1, 1, 3],  # Owner and creator are on the same node.
+                [3, 2, 3],  # Owner and borrower are on the same node.
+                [1, 3, 3],  # Creator and borrower are on the same node.
+                [3, 3, 3],  # All of them are on the same node.
+            ]
+    ])
+def test_owner_assign_when_put(ray_start_cluster, actor_resources):
+    cluster_node_config = [{
+        "num_cpus": 1,
+        "resources": {
+            f"node{i+1}": 10
+        }
+    } for i in range(3)]
+    cluster = ray_start_cluster
+    for kwargs in cluster_node_config:
+        cluster.add_node(**kwargs)
+    ray.init(address=cluster.address)
+
+    @ray.remote(resources=actor_resources["creator"], num_cpus=0)
+    class Creator:
+        def gen_object_ref(self, data="test", owner=None):
+            return ray.put(data, _owner=owner)
+
+    @ray.remote(resources=actor_resources["owner"], num_cpus=0)
+    class Owner:
+        def __init__(self):
+            self.ref = None
+
+        def set_object_ref(self, ref):
+            self.ref = ref
+
+        def warmup(self):
+            return 0
+
+    @ray.remote(resources=actor_resources["borrower"], num_cpus=0)
+    class Borrower:
+        def get_object(self, ref):
+            return ray.get(ref)
+
+    owner = Owner.remote()
+    creator = Creator.remote()
+    borrower = Borrower.remote()
+
+    # Make sure the owner actor is alive.
+    ray.get(owner.warmup.remote())
+
+    object_ref = creator.gen_object_ref.remote(data="test1", owner=owner)
+    # TODO(Catch-Bull): Ideally, deleting this line can also work normally,
+    # cause driver keep a reference of the object. But, for now, it still
+    # requires the owner to keep a reference of the object to make it
+    # available.
+    ray.get(owner.set_object_ref.remote(object_ref))
```

Comment:
In this PR, If the user wants to get an object by `ray.get` after the `creator` died, they should make sure the owner has a reference to the object.

It looks like  `creator` doesn't tell `owner` the object borrowing to `borrower`, cause `creator` is not the owner of the object (In function [CoreWorker::AllocateReturnObject](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L2043)) 

But this seems to be intuitive, this can be seen as `owner` created the object and send it to `creator`, and cleared its own reference， just like:
```python
import ray
ray.init()
@ray.remote
class A:
    def __init__(self):
        self.ref=None
    def x(self, b):
        aa = ray.put("test")
        print(aa)
        #self.ref = aa
        b.x.remote([aa])

@ray.remote
class B:
    def __init__(self):
        self.ref = None
    def x(self, x=None):
        if x:
            self.ref = x
            return x
        return self.ref
@ray.remote
class C:
    def x(self, x):
        return x
a = A.remote(); b = B.remote(); c = C.remote();
ray.get(a.x.remote(b))
data = ray.get(c.x.remote(b.x.remote()))
print("res: ", ray.get(data))
ray.kill(b)
from time import sleep; sleep(10)
print("res: ", ray.get(data))
```
and it will got `ObjectLostError` after `b` be killed.

Reply:
This example doesn't container a put with _owner=, is that a mistake?

Reply:
I just want to use this example to illustrate, `creator` got killed and the `owner` doesn't get the `object_ref`， the object be evicted. This seems reasonable.